### PR TITLE
Generalize mesa-demos and updated opengl package

### PR DESCRIPTION
**Pull request by stephencrowell**
- Changed mesa-demos to no longer depend on ```mesa``` directly. Now depends on ```gl``` with ```osmesa``` and ```glx``` variants.
- Changed ```provides('gl@:x.x', when='@x.x:)``` to ```provides('gl@:x.x', when='@x.x)``` for all versions except most recent. The former version would not be solvable because of conflicting variants.

### CODE REVIEW

File: var/spack/repos/builtin/packages/mesa-demos/package.py [hunk 1]
```diff
@@ -3,26 +3,62 @@
...
+import sys
 from spack import *
 
 
 class MesaDemos(AutotoolsPackage):
     """This package provides some demo applications for testing Mesa."""
 
     homepage = "https://www.mesa3d.org"
-    url      = "https://github.com/freedesktop/mesa-demos/archive/mesa-demos-8.3.0.tar.gz"
+    url      = "https://github.com/freedesktop/mesa-demos/archive/mesa-demos-8.3.0.tar.gz" # Should this be from free desktop gitlab?
```
Comment (chuckatkins) [inline var/spack/repos/builtin/packages/mesa-demos/package.py]: I think this is fine to leave on gh.  Just drop the associated comment.

File: var/spack/repos/builtin/packages/mesa-demos/package.py [hunk 2]
```diff
@@ -3,26 +3,61 @@
...
+
 from spack import *
 
 
 class MesaDemos(AutotoolsPackage):
     """This package provides some demo applications for testing Mesa."""
 
     homepage = "https://www.mesa3d.org"
-    url      = "https://github.com/freedesktop/mesa-demos/archive/mesa-demos-8.3.0.tar.gz"
+    url      = "https://github.com/freedesktop/mesa-demos/archive/mesa-demos-8.3.0.tar.gz"  # Should this be from free desktop gitlab?
```
Comment (chuckatkins) [inline var/spack/repos/builtin/packages/mesa-demos/package.py]: Development is essentially halted so it's last home of GH is probably correct so the comment can be dropped.

File: var/spack/repos/builtin/packages/mesa-demos/package.py [hunk 3]
```diff
@@ -3,26 +3,61 @@
...
 
     version('8.3.0', sha256='9bc1b37f4fc7bfc3f818f2d3851ffde28e8167ef11dca87f4781e9ef6206901f')
     version('8.2.0', sha256='5a9f71b815d968d0c3b77edfcc3782d0211f8520b00da9e554ccfed80c8889f6')
     version('8.1.0', sha256='cc5826105355830208c90047fc38c5b09fa3ab0045366e7e859104935b00b76d')
 
+    variant('osmesa', default=True, description="Enable the OSMesa frontend.")
+
+    is_linux = sys.platform.startswith('linux')
+    variant('glx', default=is_linux, description="Enable the GLX frontend.")
```
Comment (chuckatkins) [inline var/spack/repos/builtin/packages/mesa-demos/package.py:24]: :+1:

File: var/spack/repos/builtin/packages/mesa-demos/package.py [hunk 4]
```diff
@@ -3,26 +3,61 @@
...
     depends_on('pkgconfig', type='build')
-    depends_on('mesa')
+
+    depends_on('gl')
+    depends_on('glx', when='+glx')
+    depends_on('osmesa', when='+osmesa')
+
     depends_on('mesa-glu')
     depends_on('freetype')
     depends_on('freeglut')
```
Comment (chuckatkins) [inline var/spack/repos/builtin/packages/mesa-demos/package.py:39]: We can drop the freetype dependency

File: var/spack/repos/builtin/packages/mesa-demos/package.py [hunk 5]
```diff
@@ -3,26 +3,61 @@
...
+
     depends_on('autoconf',  type='build')
     depends_on('automake',  type='build')
     depends_on('libtool',   type='build')
     depends_on('m4',        type='build')
     depends_on('pkgconfig', type='build')
-    depends_on('mesa')
+
+    depends_on('gl')
+    depends_on('glx', when='+glx')
```
Comment (chuckatkins) [inline var/spack/repos/builtin/packages/mesa-demos/package.py:33]: I believe there is also a dependency on `libx11` and `libxext` when `+glx` is enabled

**Review by chuckatkins — CHANGES_REQUESTED** [review]

**Review by chuckatkins — DISMISSED** [review]

**Comment by chuckatkins** [comment]
@scottwittenburg looks like the e4s cleanup job is failing.  Anything to be done on our end here?

**Comment by scottwittenburg** [comment]
> @scottwittenburg looks like the e4s cleanup job is failing. Anything to be done on our end here?

We rotate the access credentials used for writing to the S3 buckets, so while I'm not sure, my hypothesis is that particular job was waiting long enough to run that the credentials expired out from under it.  If we ask spackbot to rerun, the pipeline should be provided the latest credentials and hopefully the same thing doesn't happen again.  Things have been slow in gitlab due to the `e4s-pr-generate` job generating more jobs than needed (in turn due to a bug keeping that mirror index from getting updated).

**Comment by scottwittenburg** [comment]
@spackbot run pipeline

**Comment by chuckatkins** [comment]
@spackbot run pipeline

**Comment by chuckatkins** [comment]
@scottwittenburg Looks like the e4s pipeline is still having issues, this time in generate:
```
ERROR: Job failed (system failure): error dialing backend: dial tcp 192.168.49.125:10250: connect: connection refused
```

**Review by chuckatkins — DISMISSED** [review]

**Comment by luigi-calori** [comment]
First, thanks a lot for fixing opengl package version range, it is useful also for other opengl based packages such as paraview
Furthermore, I would suggest to get rid of mesa-glu dependency as well and substituting with 
depends_on('glu')
In that case, the same thing that has been done on opengl recipe must be done on openglu package.py

**Comment by alalazo** [comment]
@stephencrowell Are you still interested in this PR? (I'm going through old PRs, and saw this one was set to be merged automatically, but wasn't merged in the end)

**Comment by alalazo** [comment]
Closing, since there has been no activity in a while. Feel free to reopen if you want to continue work on this.

**Pull request closed without merging**